sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/handler/ReceiverDeliveryHandler.java

```diff
@@ -0,0 +1,468 @@
+     */
+    public void close(String errorMessage) {
+        isTerminated.set(true);
+        messages.emitComplete((signalType, emitResult) -> {
+            logger.atVerbose()
+                .addKeyValue(ENTITY_PATH_KEY, entityPath)
+                .addKeyValue(LINK_NAME_KEY, receiveLinkName)
+                .addKeyValue(EMIT_RESULT_KEY, emitResult)
+                .log(errorMessage);
+            return false;
+        });
+    }
+
+    /**
+     * Check if the delivery received from ProtonJ is partial or already settled and log such an event.
+     *
+     * @param delivery the delivery to inspect.
+     * @return {@code true} if the delivery is partial or already settled, {@code false} otherwise.
+     */
+    private boolean isPartialOrSettledDelivery(Delivery delivery) {
+        if (delivery.isPartial()) {
+            // A message may span across Transfer frames (e.g., 200kb message over 4 Transfer
+            // frames 64k 64k 64k 8k). The ProtonJ Reactor thread will deliver each such Transfer
+            // frame as partial Delivery, then a final non-partial Delivery containing the complete
+            // message, which means the DeliveryHandler can skip any partial Deliveries.
+            final Link link = delivery.getLink();
+            if (link != null) {
+                final ErrorCondition condition = link.getRemoteCondition();
+                addErrorCondition(logger.atVerbose(), condition)
+                    .addKeyValue(ENTITY_PATH_KEY, entityPath)
+                    .addKeyValue(LINK_NAME_KEY, receiveLinkName)
+                    .addKeyValue("updatedLinkCredit", link.getCredit())
+                    .addKeyValue("remoteCredit", link.getRemoteCredit())
+                    .addKeyValue("delivery.isPartial", true)
+                    .addKeyValue("delivery.isSettled", delivery.isSettled())
+                    .log("onDelivery.");
+            } else {
+                logger.atWarning()
+                    .addKeyValue(ENTITY_PATH_KEY, entityPath)
+                    .addKeyValue("delivery.isSettled", true)
+                    .log("Partial delivery with no link.");
+            }
+            return true;
+        }
+
+        if (delivery.isSettled()) {
+            // We ran into a case where the ProtonJ Reactor thread delivered duplicate Delivery, which was
+            // settled earlier; when handling such a Delivery, ProtonJ hits an IllegalStateException.
+            // Until it is fixed in ProtonJ, DeliveryHandler needs to skip such deliveries as a workaround.
+            //
+            final Link link = delivery.getLink();
+            if (link != null) {
+                addErrorCondition(logger.atInfo(), link.getRemoteCondition())
+                    .addKeyValue(ENTITY_PATH_KEY, entityPath)
+                    .addKeyValue(LINK_NAME_KEY, receiveLinkName)
+                    .addKeyValue("updatedLinkCredit", link.getCredit())
```

Comment:
Since we have all these logging keys, maybe consider putting them as a constant in the same class? I've found that misspellings end up with keys that don't correlate to other events since we Ctrl+F everything.

Reply:
great catch, I remember the correlation issue due to misspell. I've moved these to const file.